evm_arithmetization/src/fixed_recursive_verifier.rs

```diff
@@ -1440,8 +1659,278 @@ where
+        // [`add_verifier_data_public_inputs`].
+        let count_public_inputs = builder.num_public_inputs();
+
+        let lhs = Self::add_two_to_one_block_child(&mut builder, block_circuit);
+        let rhs = Self::add_two_to_one_block_child(&mut builder, block_circuit);
+
+        let lhs_public_values = lhs.public_values(&mut builder);
+        let rhs_public_values = rhs.public_values(&mut builder);
+
+        let lhs_pv_hash = builder.hash_n_to_hash_no_pad::<C::InnerHasher>(lhs_public_values);
+        let rhs_pv_hash = builder.hash_n_to_hash_no_pad::<C::InnerHasher>(rhs_public_values);
+        let mut mix_vec = vec![];
+        mix_vec.extend(&lhs_pv_hash.elements);
+        mix_vec.extend(&rhs_pv_hash.elements);
+        let mix_pv_hash = builder.hash_n_to_hash_no_pad::<C::InnerHasher>(mix_vec);
+
+        debug_assert_eq!(
+            count_public_inputs,
+            builder.num_public_inputs(),
+            "Public inputs were registered after called `add_verifier_data_public_inputs`"
+        );
+
+        debug_assert_eq!(
+            block_circuit.circuit.common.num_public_inputs,
+            builder.num_public_inputs(),
+            "The block aggregation circuit and the block circuit must agree on the number of public inputs."
+        );
+
+        let circuit = builder.build::<C>();
+        TwoToOneBlockCircuitData {
+            circuit,
+            lhs,
+            rhs,
+            mix_pv_hash,
+            dummy_pis,
+            cyclic_vk,
+        }
+    }
+
+    /// Used in the case of a non aggregation transaction child.
+    /// Creates dummy public inputs to set the cyclic vk to the aggregation
+    /// circuit values, so that both aggregation and non-aggregation parts
+    /// of the child share the same vk. This is possible because only the
+    /// aggregation inner circuit is checked against its vk.
+    fn set_dummy_proof_with_cyclic_vk_pis(
+        circuit_agg: &CircuitData<F, C, D>,
+        witness: &mut PartialWitness<F>,
+        agg_proof: &ProofWithPublicInputsTarget<D>,
+        proof: &ProofWithPublicInputs<F, C, D>,
+    ) {
+        let ProofWithPublicInputs {
+            proof,
+            public_inputs,
+        } = proof;
+        let ProofWithPublicInputsTarget {
+            proof: proof_targets,
+            public_inputs: pi_targets,
+        } = agg_proof;
+
+        // The proof remains the same.
+        witness.set_proof_target(proof_targets, proof);
+
+        let num_pis = circuit_agg.common.num_public_inputs;
+        let mut dummy_pis = vec![F::ZERO; num_pis];
+        let cyclic_verifying_data = &circuit_agg.verifier_only;
+        let mut cyclic_vk = cyclic_verifying_data.circuit_digest.to_vec();
+        cyclic_vk.append(&mut cyclic_verifying_data.constants_sigmas_cap.flatten());
+
+        let cyclic_vk_len = cyclic_vk.len();
+        for i in 0..cyclic_vk_len {
+            dummy_pis[num_pis - cyclic_vk_len + i] = cyclic_vk[i];
+        }
```

Comment:
You may want to add a one-liner explanation as to why we write the VK at the end of the dummy PIs.

Reply:
(I don't think we need to explain why `num_pis > cyclic_vk_len`, all these circuits encode the VK in their PIs anyway).

Reply:
I have added a one-liner, reworked the comments, and renamed parameters to avoid shadowing; e.g `proof` was used for different things.

nit math: I think it is possible to have `num_pis = cyclic_vk_len`.